fix(ui-markdown-editor): linebreak - #263, #267, #269

d-e-v-esh opened the pull request:
<!--- Provide a formatted commit message describing this PR in the Title above -->
<!--- See our DEVELOPERS guide below: -->
<!--- https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format -->
# Closes #263 #267
<!--- Provide an overall summary of the pull request -->

## Update
**_This was my thought process of finding the problem and the solution._**

## Identifying the problem

 - As you can see [here](https://github.com/accordproject/web-components/blob/b6ccb71ca4e103d5c699ad4395653d55597f3919/packages/ui-markdown-editor/src/utilities/toolbarHelpers.js#L98), we do not have any function for the ``softbreak``.
 - The functions that exist there are ``insertLinebreak`` and ``insertHeadingbreak``. 



 - So I thought that maybe ``softbreak`` was not working because there is no function for it. But that is actually not the case. 
 - The ``softbreak`` was not working because of this conditional statement [here](https://github.com/accordproject/web-components/blob/b6ccb71ca4e103d5c699ad4395653d55597f3919/packages/ui-markdown-editor/src/index.js#L93). This basically means that if ``Enter`` is being pressed from a non-heading block then we totally skip out hotkey actions. That is also the reason why pressing `Ctrl` + `Enter` did not inset a page break. **_If you remove that statement then `softbreak` and page break works like normal._**


 - And this should TOTALLY prevent us from getting a `linebreak` in a paragraph. And the only reason pressing `Enter` is still functional is because of [this](https://github.com/accordproject/web-components/blob/b6ccb71ca4e103d5c699ad4395653d55597f3919/packages/ui-markdown-editor/src/plugins/withLists.js#L9) and [this](https://github.com/accordproject/web-components/blob/b6ccb71ca4e103d5c699ad4395653d55597f3919/packages/ui-markdown-editor/src/plugins/withLinks.js#L66) functions where we are overriding ``insertBreak()``. You can put a few ``console.log()`` there and check which functions work and which do not.


 - That is also the reason why inserting a page break (`Ctrl` + `Enter`) and `softbreak`(`Shift` +  `Enter`) works absolutely fine for all the heading blocks because this conditional statement only restricts non-heading blocks from all the `Enter` functions.


## Exchange in types of `break` **Bug**
 - As I mentioned earlier in the second point that the only functions that existed are the ``insertLinebreak`` and ``insertHeadingbreak``.
 - To find out which function is doing what, we need to remove this conditional statement [here](https://github.com/accordproject/web-components/blob/b6ccb71ca4e103d5c699ad4395653d55597f3919/packages/ui-markdown-editor/src/index.js#L93) and simply add a `console.log()` statement after [this](https://github.com/accordproject/web-components/blob/b6ccb71ca4e103d5c699ad4395653d55597f3919/packages/ui-markdown-editor/src/index.js#L101) line like this:

```
    hotkeys.forEach((hotkey) => {
      if (isHotkey(hotkey, event)) {
        event.preventDefault();
        const { code, type } = HOTKEYS[hotkey];
        console.log(code) <===
        hotkeyActions[type](code);
      }
    })
```
By doing this, we get:

![VoIURQSy9U](https://user-images.githubusercontent.com/59534570/110131027-bad35580-7def-11eb-9345-0e590c6553a3.gif)


### Here:
Pressing `Enter` results in a => `headingbreak`
Pressing `Shift` + `Enter` results in a => `linebreak`

This is the case because this is how it is mentioned in the ``hotkeys`` file as you can see [here](https://github.com/accordproject/web-components/blob/b6ccb71ca4e103d5c699ad4395653d55597f3919/packages/ui-markdown-editor/src/utilities/hotkeys.js#L52).
 - And now we know that pressing `Shift` + `Enter` is creating a `softbreak` as it should but it is named as a `linebreak` and
- Pressing `Enter` is always creating a false `headingbreak` no matter the type of block. It is false `headingbreak` and does not do anything because of the problem I mentioned above that is pressing `Enter` is useless in a non-heading block because of this conditional statement [here](https://github.com/accordproject/web-components/blob/b6ccb71ca4e103d5c699ad4395653d55597f3919/packages/ui-markdown-editor/src/index.js#L93)

---

What is happening till now:
Every time we press Enter, a heading break is going to get inserted.

How it was implemented was:
1. We created a boolean which tells us if the current block is a heading.
2. We created a restriction on the `enter` keyword when being pressed from a non-heading block. This was causing #263 
3. If enter is called from a `headingblock` then we will go to our custom hotkeyActions and create a heading break
4. But if the `enter` is called from a non - heading block then we will skip our custom hotkeyActions and a default enter linebreak will occur



### Changes
<!--- More detailed and granular description of changes -->
<!--- These should likely be gathered from commit message summaries -->

1. Converted `headingbreak` to linebreak
2. This converts every single enter press to a simple linebreak and if enter is pressed from a heading block, only then we will make it a `headingbreak`
3. This removes the restriction on the enter key in the paragraph so the page break works.


### Related Issues
- Issue  #263 #267 #269 
- Pull Request #220 

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [x] Vital features and changes captured in unit and/or integration tests
- [ ] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [x] Merging to `master` from `fork:branchname`
- [x] Manual accessibility test performed
    - [ ] Keyboard-only access, including forms
    - [ ] Contrast at least WCAG Level A
    - [ ] Appropriate labels, alt text, and instructions

d-e-v-esh committed:
fix(ui-markdown-editor): fixes cursor movement around headings - #260

Signed-off-by: d-e-v-esh <59534570+d-e-v-esh@users.noreply.github.com>

jolanglinais reviewed (changes requested):
We'll need to do a thorough investigation on all of this to see what is going on, the limitations, and what these changes do.

d-e-v-esh commented:
@irmerk That is great. Will that take very long?

dselman commented on packages/ui-markdown-editor/src/components/index.js:
```diff
@@ -58,6 +57,7 @@ const Element = (props) => {
     }
   };
   const elementRenderer = customElements
+    // console.log(customElements)
```
Remove please.

dselman commented on packages/ui-markdown-editor/src/index.js:
```diff
@@ -29,6 +27,9 @@ export const markdownToSlate = (markdown) => {
 };
 
 export const MarkdownEditor = (props) => {
+
```
Remove please

dselman commented on packages/ui-markdown-editor/src/index.js:
```diff
@@ -90,15 +91,17 @@ export const MarkdownEditor = (props) => {
       if (isHotkey(hotkey, event)) {
         event.preventDefault();
         const { code, type } = HOTKEYS[hotkey];
+        // if linebreak happens form a heading block then we run the insertHeadingBrake function
```
Typo in xxxBrake

dselman commented on packages/ui-markdown-editor/src/index.js:
```diff
@@ -90,15 +91,17 @@ export const MarkdownEditor = (props) => {
+        // if linebreak happens form a heading block then we run the insertHeadingBrake function
+        // other it is a linebreak every time
+        if(type === 'linebreak' && isBlockHeading(editor)){
+          hotkeyActions['headingbreak']()
```
Use the 'headingbreak' constant please.

dselman commented on packages/ui-markdown-editor/src/utilities/toolbarHelpers.js:
```diff
@@ -66,6 +64,7 @@ export const toggleMark = (editor, format) => {
 };
 
 export const toggleHistory = (editor, format) => {
+  console.log(format)
```
Remove please.

dselman commented on packages/ui-markdown-editor/src/utilities/toolbarHelpers.js:
```diff
@@ -91,17 +90,26 @@ export const insertThematicBreak = (editor, type) => {
-export const insertLinebreak = (editor, type) => {
+export const insertSoftBreak = (editor, type) => {
   const text = { object: 'text', text: '' };
   const br = { type, children: [text], data: {} };
```
Why take `type` as an argument if this is just for soft break?

irmerk commented:
> @irmerk That is great. Will that take very long? Should I write the description of changes and limitations in more detail?

Yeah I think so. This is a sensitive change, so as much helpful context would be great.

d-e-v-esh commented:
@dselman @irmerk I have updated the pull-request body and added more context highlighting the problem. Does this bring more clarity?

dselman commented:
When I test this using the Netlify preview there's something strange happening with the selection. Try selecting multiple paragraphs forwards, or backwards, and you see that the selection is getting extended outside of the selection target.

d-e-v-esh commented:
@dselman I don't see that behavior. Can you make a screen recording to show exactly what is strange?

K-Kumar-01 commented:
@d-e-v-esh 
https://www.loom.com/share/56a2f8833f9f481e96dd5582fce0b745
I think @dselman may be talking about this thing.

dselman commented:
> @dselman I don't see that behavior. Can you make a screen recording to show exactly what is strange?

It seems to be specific to Safari - I don't see it on Chrome.

d-e-v-esh commented:
@K-Kumar-01 Is this happening only on this update or is this on the master branch? You cut out the URL.

K-Kumar-01 commented:
> @K-Kumar-01 Is this happening only on this update or is this on the master branch? You cut out the URL.

@d-e-v-esh 
It is happening in the deployment of this PR only.
I have rechecked it. 
I clicked on the deploy preview and then made the above video.

d-e-v-esh commented:
@K-Kumar-01 That is actually a really bad bug. I'll see what is causing the problem and update.

jolanglinais reviewed (changes requested):
We still haven't found time yet, but this needs investigation by @accordproject/maintainers-ui.

In the meantime @d-e-v-esh you could rebase all your commits to clean them up and put DCO signoffs on them.

d-e-v-esh commented:
@irmerk Does this look good now?

irmerk commented:
Yes looks good. I think wait to rebase again (because there are conflicting files after I merged a couple other PRs) until we review.

irmerk commented:
I've finally been able to look more into this. This is really thorough and I think is on the right track. Some residual awkward behavior I'm seeing is when you press <kbd>ENTER</kbd> at the end of "**My Heading**" at the top of the [demo](https://deploy-preview-270--ap-web-components.netlify.app/?path=/story/markdown-editor--demo), it inserts two new paragraphs, both of which are not headings, and places the cursor at the beginning of the second one. Could this be because we are inserting `lineBreakNode` in `insertLineBreak`?

irmerk commented on packages/ui-markdown-editor/src/utilities/toolbarHelpers.js:
```diff
@@ -91,17 +89,26 @@ export const insertThematicBreak = (editor, type) => {
   const text = { object: 'text', text: '' };
   const n = { object: "block", type: 'paragraph', children: [text] };
   Transforms.insertNodes(editor, n);
+  Transforms.move(editor, { distance: 1, unit: 'character' })
```
I ran this locally and actually I think it could be this line that we may want to back out.

irmerk commented:
As of right now #263 and #269 are solved for sure. I _believe_ #267 is resolved, will want extra eyes on this from @dselman and/or @jeromesimeon.

d-e-v-esh commented:
@irmerk Oh yes, I remember that being very annoying. I don't know why I kept it, probably because it was there from the beginning. And we were talking about implementing [this](https://github.com/accordproject/web-components/issues/260#issuecomment-790687176) after the linebreak issue is resolved. It is fixed now.

irmerk commented:
Great! It looks like there are still merge conflicts, so you'll want to rebase to resolve those.

d-e-v-esh commented:
@irmerk I am not sure if this worked. Did the conflicts go away? I can't see them.

irmerk commented:
Suspicious, I refreshed and it said the conflicts were there, refreshed again and they were gone.

irmerk commented:
There are still two issues with headings:

1. Place cursor at beginning of "My Heading"
Press <kbd>ENTER</kbd>
See that new normal line is created before and cursor is placed above "My Heading"
2. Place cursor in the middle of "My Heading"
Press <kbd>ENTER</kbd>
See that new normal line is created in between "My Heading" where cursor was placed.

irmerk commented:
Note: I'm looking at your code locally to see if I can see the solution.

d-e-v-esh commented:
@irmerk  I think we can create a new and fully working headingBreak with all the bugs fixed under the PR of #260 after this is merged. I think that would be better, isn't it?

irmerk commented:
@d-e-v-esh we're going to put the changes in this PR on the working group call tomorrow to try to make sure it doesn’t lead to further confusion around linebreaks/softbreaks. If you’ll be on during the call, feel free to add to the conversation. If not, I think @dselman and I can present what the changes will do.

dselman reviewed (changes requested):
This looks very clean - however when I try the Netlify preview and I forward-select past a line break something strange happens with the selection (whole blocks get selected). I can't reproduce the same issue with the master: https://ap-web-components.netlify.app/?path=/story/markdown-editor--demo

d-e-v-esh commented:
@dselman I think the reason for that seems to be [this](https://github.com/accordproject/web-components/pull/270#issuecomment-794986866) that I talked about earlier. It seems like something changed in the conversion. That behavior is not present in the master and it was also not present when I made the PR. Do you know where it could be coming from?

dselman commented:
> @dselman I think the reason for that seems to be [this](https://github.com/accordproject/web-components/pull/270#issuecomment-794986866) that I talked about earlier. It seems like something changed in the conversion. That behavior is not present in the master and it was also not present when I made the PR. Do you know where it could be coming from?

Sorry, no idea. Can you reproduce the issue when you run locally using this branch?

d-e-v-esh commented:
@dselman Yes I can. It doesn't go back to when this was not happening.

irmerk commented:
Update: This PR should no longer address #267

irmerk commented:
Also, this should unblock #260 and #284

d-e-v-esh commented:
@irmerk I have implemented the new `paragraphBreak` and also fixed the error that @dselman pointed out. I think it should be perfect now. Now we just need to update the demo text on the editor.

jolanglinais reviewed (approved):
Yes, this looks good to me! Will want @dselman on this as well.

irmerk commented:
Pinging @dselman on this again.

dselman commented:
@d-e-v-esh looks great. Can you please resolve the merge conflicts and then I will merge.

dselman reviewed (changes requested):
Looks great. Please resolve the merge conflicts and I will merge.

dselman commented:
Would love to get this into the next release (this week). @d-e-v-esh will you have time to resolve the merge conflicts?

d-e-v-esh committed:
fix(ui-markdown-editor): linebreak - #263, #267

Signed-off-by: Devesh <deves125@gmail.com>

d-e-v-esh commented:
@dselman I removed all the merge conflicts and it works totally great. You can check it out.

Michael-Grover reviewed (approved):
lgtm, thanks @d-e-v-esh !
